doc/ps_design.md

```diff
@@ -1,19 +1,80 @@
+Each worker has a local copy of the model variables. After the master relaunches a PS, the PS pod can recover model variables from workers. For embedding vectors, PS must create replicas to support fault tolerance. For each PS *P(i)*, it will store *M* replicas in the following *M* PSs from *P(i+1 % N)* to *P(i+M % N)*. The relaunched PS can recover embedding vectors from one of its replicas. If there are more than *M* continuously-indexed PSs failing, at least one PS fails with all of its replica PSs. The ElasticDL job has to recover from a recent checkpoint.
 
 ## Parameter Server (PS)
 
 ## Interactions among Master, PS and Worker
-### Master - PS
-### Master - Worker
-### Worker - PS
+
 
 ## Embedding Replicas in PS
+An ElasticDL job has *N* PSs. Embedding vectors are partitioned into these *N* PSs. The user provides *M*, the number of replicas for embedding vectors. *M* must be smaller than *N* as each PS uses other PSs to store its embedding replicas.
+
+Assume *E(i)* is the embedding vectors in PS *PS(i)*, it has *M* replicas which are stored in PSs from *P(i + 1 % N)* to *P(i + M % N)*. Also, *PS(i)* has replicas for *E(i - M % N)* to *E(i - 1 % N)*. 
+
+*PS(i)* stores *E(i)* as a dictionary so that every embedding vector has a corresponding key.  *PS(i)* maintains *M* updated embedding vector key sets *UKS_i(j) for j from 0 to M - 1*. When *PS(i)* sparsely updates its embedding vectors *E(i)*, it also add the updated embedding vector keys into these *M* sets. 
+
+
+*PS(i)* also periodically synchronize the replicas stored in it from *E(i - M % N)* to *E(i - 1 % N)*. The synchronization frequency can be several seconds.
+
+Each PS will provide a GRPC service for the replica synchronization.
+
+```
+message SynchronizeEmbeddingRequest {
+    int32 replica_index = 1;
+}
+
+message SynchronizeEmbeddingResponse {
+    map<string, Tensor> embedding_vectors = 1;
+}
+
+# GRPC service for replica synchronization
+rpc SynchronizeEmbedding(SynchronizeEmbeddingRequest) returns (SynchronizeEmbeddingResponse);
+
+# GRPC service for PS to recover embedding vectors after relaunch
+rpc GetReplica(SynchronizeEmbeddingRequest) returns (SynchronizeEmbeddingResponse);
+```
+
+Each PS has a thread dedicated for replica synchronization. In this dedicated thread, PS will synchronize the stored replicas in it.
+
+```
+# T is the number of seconds for synchronization frequency
+# Assume current PS is PS(i), self._stub[index] is the stub for PS(i - index)'s GRPC server.
+# self.replicas[index] is the replica for PS(i - index).
```

Comment:
Is the range of `index` [0,M) or [1, M]?
If the range is [0, M), `self.replicas[index]` should be the replica for PS(i - index - 1).